### PR TITLE
make `ChromiumDevice.stopApp` a no-op if it has already been called

### LINKED ISSUE #156700: When closing Chrome at the end of `flutter run --machine`: `WebSocketException: Connection to 'http://localhost55388/devtools/page/8B8BA1FC234B2A5EBF0770CE0DA9255E#' was not upgraded to websocket`
Since 3.24.0

Command: `flutter run` (mostly with `--machine`)

Represents about 1% on crashes on 3.24.3 (which has ~1% of users experiencing crashes)

There are a couple of other stack traces, but 96% of them looked like this:

```
WebSocketException: WebSocketException: Connection to 'http://localhost:55388/devtools/page/8B8BA1FC234B2A5EBF0770CE0DA9255E#' was not upgraded to websocket
at _WebSocketImpl.connect(websocket_impl.dart:1011)
at WebSocket.connect(websocket.dart:320)
at WipConnection.connect(webkit_inspection_protocol.dart:231)
at ChromeTab.connect(webkit_inspection_protocol.dart:184)
at Chromium.close(chrome.dart:515)
at <asynchronous gap>(async)
at ChromiumDevice.stopApp(web_device.dart:164)
at <asynchronous gap>(async)
at ResidentWebRunner._cleanup(resident_web_runner.dart:203)
at <asynchronous gap>(async)
at ResidentWebRunner.cleanupAtFinish(resident_web_runner.dart:191)
at <asynchronous gap>(async)
at ResidentRunner.waitForAppToFinish(resident_runner.dart:1482)
at <asynchronous gap>(async)
at RunCommand.runCommand(run.dart:788)
at <asynchronous gap>(async)
at FlutterCommand.run.<anonymous closure>(flutter_command.dart:1408)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at CommandRunner.runCommand(command_runner.dart:212)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand.<anonymous closure>(flutter_command_runner.dart:420)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand(flutter_command_runner.dart:364)
at <asynchronous gap>(async)
at run.<anonymous closure>.<anonymous closure>(runner.dart:130)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:153)
at <asynchronous gap>(async)
at main(executable.dart:93)
at <asynchronous gap>(async)
```

### PR DESCRIPTION
**Pull request by andrewkolos**
(Hopefully) fixes https://github.com/flutter/flutter/issues/156700

To summarize that thread, there are a few ways to trigger cleanup logic in `flutter run -d chrome`. For example, pressing `q`  will close Chrome, but Chrome closing triggers cleanup logic in the resident runner (which includes trying to close Chrome, which is redundant). 

A simple way to fix this is to not call `Chrome.close` in `ChromiumDevice.stopApp` when we've already called it once. This solution isn't ideal—consider a scenario where two calls to `stopApp` are made concurrently. The second one would probably complete before the browser is actually closed. However, I don't foresee anything ever depending on `stopApp` finishing implying that the browser is closed.

I feel like there might be a yak to shave here (`ResidentWebRunner` could keep track of its state and not initiate redundant cleanup operations), but I am not sure its worth the continued effort.

<details>

<summary> Pre-launch checklist </summary> 

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

</details>

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/test/web.shard/chrome_test.dart [hunk 1]
```diff
@@ -907,6 +907,35 @@ void main() {
     expect(await chromiumLauncher.connect(chrome, false), equals(chrome));
     await chrome.close();
   });
+
+  // testWithoutContext('close is a no-op if it has already been called', () async {
```
Comment (bkonyi) [inline packages/flutter_tools/test/web.shard/chrome_test.dart:911]: Do we still need this code?

File: packages/flutter_tools/test/general.shard/web/devices_test.dart [hunk 2]
```diff
@@ -430,3 +459,36 @@ class TestChromiumLauncher implements ChromiumLauncher {
...
+  Future<void> close() async {
+    if (_closed) {
+      throw Exception('Already closed');
+    }
+    _closed = true;
+  }
+}
+
+class _UnimplementedChromium extends Fake implements Chromium {
+
```
Comment (bkonyi) [inline packages/flutter_tools/test/general.shard/web/devices_test.dart:493]: Nit: just keep the empty body on the same line as the class declaration.

**Review by bkonyi — APPROVED** [review]
LGTM w/ nits

File: packages/flutter_tools/test/web.shard/chrome_test.dart (hunk 1 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/web.shard/chrome_test.dart:911]: Oops, this was from a previous attempt at fixing this. Thanks for catching!

**Pull request merged**